examples/reference/custom_components/ReactiveESM.md

```diff
@@ -0,0 +1,283 @@
+
+#### `render` Function
+
+The `render` function can accept the following arguments:
+
+- **`children`**: DON'T KNOW HOW TO EXPLAIN
+- **`data`**: The `data` element corresponds to the custom components Python Parameters and provides methods to `.watch` for parameter changes and `.send_event` back to Python.
+- **`el`**: The main html element of the component. Append elements to this element to display them. IS THIS OBSOLETE AS YOU SHOULD BE RETURNING THE ELEMENT TO RENDER?
+- **`html`**: The `html` function enables JSX-like syntax in plain JavaScript without a transpiler. See [`htm`](https://github.com/developit/htm). IS THIS OBSOLETE?
+- **`model`**: The Bokeh model. DON'T KNOW HOW TO EXPLAIN
+- **`view`**: The Bokeh view. DON'T KNOW HOW TO EXPLAIN
+
+The `render` function should return the html element to display.
+
+#### Lifecycle Methods
+
+The `render` function is run for each independent instance of the component. Besided the `render` function you can also export
+
+- `initialize`: Is executed once per widget instance, during model initialization. It has access to model to setup non-view event handlers or state to share across views. ANYWIDGET NEEDS THIS.
+- `after_layout`: DO WE STILL NEED SOMETHING LIKE THAT TO RESIZE OR REDRAW?
+- `teardown`: DO WE NEED SOME EVENT TO CLEAN UP?
+
+## Usage
+
+### CSS
+
+You can include CSS in the `_stylesheets` attribute. The CSS will be injected into the component.
+
+```python
+import panel as pn
+import param
+
+pn.extension()
+
+class StyledCounterButton(pn.ReactiveESM):
+
+    value = param.Integer()
+
+    _stylesheets = [
+        """
+        button {
+            background: #0072B5;
+            color: white;
+            border: 0px solid black;padding:10px;
+            border-radius:4px;
+        }
+        button:hover {
+            background: #4099da;
+        }
+        """
+    ]
+
+    _esm = """
+    export function render({ data }) {
+        let btn = document.createElement("button");
+        btn.innerHTML = `count is ${data.value}`;
+        btn.addEventListener("click", () => {
+            data.value+=1
+        });
+        data.watch(() => {
+            btn.innerHTML = `count is ${data.value}`;
+          }, 'value')
+        return btn
+    }
+    """
+
+StyledCounterButton().servable()
+```
+
+## Send Events from Javascript to Python
+
+You can send events from JavaScript to Python using the `data.send_event` method. You can handle the event in Python by defining an `on_<event_name>` method.
+
+```python
+import panel as pn
+import param
+
+pn.extension()
+
+class ButtonEventExample(pn.ReactiveESM):
+
+    value = param.Parameter()
+
+    _esm = """
+    export function render({ data }) {
+        const btn = document.createElement('button')
+        btn.innerHTML = `Click Me`;
+        btn.onclick = (event) => data.send_event('click', event)
+        return btn
+    }
+    """
+
+    def on_click(self, event):
```

Comment:
Not sure if this should be the pattern or whether it should be some private method like `_handle_<event-name>`.

Reply:
Private is probably better as it should not be exposed to users of the class.

Reply:
Agreed.

Reply:
Textual does not use `_` names though.
![image](https://github.com/holoviz/panel/assets/42288570/f07154d1-3b02-43ad-b43d-659a210c27d2)

Reply:
Guess it's a question of target audience, are you building a component for yourself or are you building a component for reuse by yourself and others?

Reply:
I would still argue `_`. Because it will work for both audiences.

Reply:
Okay, done, it's now `_handle_<event>`